zk/src/org/zkoss/zk/ui/sys/ShadowElementsCtrl.java

```diff
@@ -103,18 +104,16 @@ public static final Object getDistributedIndexInfo() {
 			}
 
 			// the following code will filter the shadow element if any.
-			Component end = shadows[length - 1];
-			if (end instanceof ShadowElementCtrl) {
-				ShadowElementCtrl se = (ShadowElementCtrl) end;
-				end = se.getLastInsertion();
-			}
-
+			// in multiple shadows case, some of nextInsertion would point to the useless component, which would detach
+			// later. So here rewrite the list add method.
 			LinkedList<Component> list = new LinkedList<Component>();
-			while (start != null) {
-				list.add(start);
-				if (start == end)
-					break;
-				start = start.getNextSibling();
+			for (Component comp : shadows) {

```

Comment:
is "start" useless now? if it is useless, then you should modify more code in this function (previous lines)